app/cdap/components/RecipeList/reducer.ts

```diff
@@ -0,0 +1,171 @@
+  recipes: IRecipe[];
+  ready: boolean;
+}
+
+const Actions = {
+  setRecipes: 'SET_RECIPES',
+  prevPage: 'RECIPE_PREV_PAGE',
+  nextPage: 'RECIPE_NEXT_PAGE',
+  setSort: 'RECIPE_SET_SORT',
+  reset: 'RECIPE_RESET',
+  setInitValues: 'RECIPE_SET_INIT_VALUES',
+};
+
+export const defaultInitialState: IState = {
+  recipes: null,
+  ready: false,
+  nextPageToken: null,
+  previousTokens: [],
+  pageToken: null,
+  pageLimit: 10,
+  sortColumn: 'name',
+  sortedOrder: SortOrder.ASCENDING,
+};
+
+export const reducer = (state: IState, action: IAction) => {
+  switch (action.type) {
+    case Actions.setInitValues:
+      return {
+        ...state,
+        sortColumn: action.payload.sortBy,
+        sortedOrder: action.payload.sortOrder,
+        pageLimit: action.payload.pageSize,
+      };
+    case Actions.setSort:
+      return {
+        ...state,
+        sortColumn: action.payload.sortColumn,
+        sortedOrder: action.payload.sortOrder,
+        pageToken: null,
+        previousTokens: [],
+        nextPageToken: null,
+        ready: false,
+      };
+    case Actions.prevPage:
+      const lastTokenIdx = state.previousTokens.length - 1;
+      const lastToken = state.previousTokens[lastTokenIdx];
+      return {
+        ...state,
+        previousTokens: state.previousTokens.slice(0, lastTokenIdx),
+        nextPageToken: state.pageToken,
+        pageToken: lastToken,
+        ready: false,
+      };
+    case Actions.nextPage:
+      return {
+        ...state,
+        previousTokens: [...state.previousTokens, state.pageToken],
+        nextPageToken: null,
+        pageToken: state.nextPageToken,
+        ready: false,
+      };
+    case Actions.setRecipes:
+      return {
+        ...state,
+        recipes: action.payload.recipes,
+        nextPageToken: action.payload.nextPageToken,
+        ready: true,
+      };
+    case Actions.reset:
+      return { ...defaultInitialState, pageLimit: state.pageLimit };
+    default:
+      return state;
+  }
+};
+
+export function reset(dispatch, state) {
+  dispatch({
+    type: Actions.reset,
+  });
+  getSavedRecipes(dispatch, { ...defaultInitialState, pageLimit: state.pageLimit });
+}
+
+export function getSavedRecipes(dispatch, state) {
+  const { pageToken, sortedOrder, sortColumn, pageLimit } = state;
+
+  MyDataPrepApi.getRecipeList({
+    context: getCurrentNamespace(),
+    pageToken,
+    sortBy: sortColumn,
+    pageSize: pageLimit,
+    sortOrder: sortedOrder,
+  }).subscribe((recipesResponse) => {
+    dispatch({
+      type: Actions.setRecipes,
+      payload: {
+        recipes: recipesResponse.values,
+        nextPageToken: recipesResponse.nextPageToken,
+      },
+    });
+  });
+}
+
+export function prevPage(dispatch, state) {
+  const { previousTokens } = state;
+  if (!previousTokens.length) {
+    return;
+  }
+  dispatch({
+    type: Actions.prevPage,
+  });
+  getSavedRecipes(dispatch, { ...state, pageToken: previousTokens[previousTokens.length - 1] });
```

Comment:
Can you combine the two dispatches into one dispatch? maybe write a `getPrevPageRecipes` function. In there you call the api with prevPageToken and dispatch

Reply:
I've tried but not able to update the state accurately. I will discuss with you to get the way to do so.

Reply:
Combined dispatches now. Please latest change set.